include/zephyr/spinlock.h

```diff
@@ -29,6 +29,35 @@ extern "C" {
  * @{
  */
 
+/**
+ * @brief Execute a code block with the given spinlock held, automatically
+ * releasing it in the end.
+ *
+ * @details Example usage:
+ *
+ * @code{.c}
+ * WITH_SPINLOCK(&mylock) {
+ *   ...execute protected statements with the lock held...
+ * }
+ * @endcode
+ *
+ * This basically expands to:
+ *
+ * @code{.c}
+ * for (k_spinlock_key_t key = k_spin_lock(&mylock); k_spin_unlock(&mylock, key); ...) {
+ *     ...execute protected statements with the lock held...
+ * }
+ * @endcode
+ *
+ * @note This pattern may only be used, if you need to protect a critical
+ * section with a single point of exit in the end. As it uses a for-loop
+ * internally, you MUST NOT EXIT ANYWHERE FROM THE BLOCK (e.g. using a break or
+ * return statement) otherwise the spinlock will not be released.
+ */
```

Comment:
There is an additional caveat that might be worth noting.  As this for-loop has only a single iteration, using a _continue_ statement may allow one to safely leave the block. This is on the surface somewhat counter intuitive, so it may be worthwhile to provide a macro that expands to _continue_ that can be used to safely leave the block.

Reply:
Good point. See whether you like my proposal.